src/vm/memory/segments.zig

```diff
@@ -204,20 +221,33 @@ pub const MemorySegmentManager = struct {
         return self.segment_used_sizes;
     }
 
-    /// Retrieves the size of a memory segment by its index if available, else computes it.
-    ///
-    /// This function attempts to retrieve the size of a memory segment by its index. If the size
-    /// is not available in the segment sizes map, it calculates the effective size and returns it.
-    ///
-    /// # Parameters
-    ///
-    /// - `index` (u32): The index of the memory segment.
+    /// Computes the first relocated address of each memory segment
     ///
+    ///  Relocation Logic:
+    ///     Step 1: Get segment sizes:
+    ///              0 --(has size)--> 3
+    ///              1 --(has size)--> 5
+    ///              2 --(has size)--> 1
+    ///     Step 2: Step 2: Assign a base to each segment:
+    ///              0 --(has base value)--> 1
+    ///              1 --(has base value)--> 4 (that is: 1 + 3)
+    ///              2 --(has base value)--> 9 (that is: 4 + 5)
     /// # Returns
     ///
-    /// A `u32` representing the size of the segment or a computed effective size if not available.
-    pub fn getSegmentSize(self: *Self, index: u32) ?u32 {
-        return self.segment_sizes.get(index) orelse self.getSegmentUsedSize(index);
+    /// A `Slice` of the `ArrayList(u32)` representing the relocated segments.
+    pub fn relocateSegments(self: *Self, allocator: Allocator) !ArrayList(usize).Slice {
+        const first_addr = 1;
+        var relocatable_table = ArrayList(usize).init(allocator);
+        errdefer relocatable_table.deinit();
+        try relocatable_table.append(first_addr);
+        for (self.segment_used_sizes.keys()) |key| {
+            const index = self.segment_used_sizes.getIndex(key) orelse return MemoryError.MissingSegmentUsedSizes;
+            const segment_size = self.getSegmentSize(@intCast(index)) orelse return MemoryError.MissingSegmentUsedSizes;
+            try relocatable_table.append(relocatable_table.items[index] + segment_size);
+        }
+        // The last value corresponds to the total amount of elements across all segments, which isnt needed for relocation.
+        _ = relocatable_table.pop();
+        return relocatable_table.toOwnedSlice();
```

Comment:
Is there a special reason why we return a slice here and not an ArrayList as we do on other functions?

Reply:
https://github.com/keep-starknet-strange/ziggy-starkdust/pull/156/#discussion_r1415557401

Reply:
Ok fine